jsgen: use the new-style CGIR

zerbina opened the pull request:
## Summary

* use the goto-using CGIR for JavaScript code generation
* as a side-effect, multiple bugs with tracking of the current
  exception are fixed 
* exceptions thrown from `asm` or `.emit` code cannot be expected to be
  caught anymore (this was already undefined behaviour)

## Details

The CGIR for a body is first passed to the new `jsflow` module, where
the shape of the necessary JavaScript is computed, with the focus on
the control-flow constructs (`if`, `switch`, `try`, etc.). Constructs
are translated as follows:
* `cnkGotoStmt` and `cnkJoinStmt` are translated to `break` and labeled
  blocks (JavaScript doesn't support goto directly)
* `cnkCaseStmt` dispatchers are translated to `switch`/`case`
  statements
* `cnkFinally` and `cnkExcept` are translated to `try`/`finally` and
  `try`/`catch`, respectively

It's first computed which statements `try`s and labeled blocks *must*
enclose. Afterwards, their starts are reordered such that valid
JavaScript is formed.

Since `finally` and `catch` clauses can then intercept `break`s or
`throw`s that they shouldn't, `finally` and `catch` clauses are guarded
by a boolean flag. Immediately before an explicit (`break` and `throw`)
or implicit (procedure call) jump that would enter a clause it
shouldn't, the boolean locals for the problematic clauses are set to
false. Upon exit of a disabled clause, the local is set back to true.

#### Case Statements

Direct translation of `cnkCaseStmt`s into `switch`/`case` statement
where each branch is a `break` would result in very deep block nesting
for `case` statements with many branches. The resulting JavaScript code
can cause the V8 JavaScript engine to fail/crash when compiling the
code!

Therefore, the jump targets need to be inlined whenever possible.
`jsflow` computes the code regions eligible for inlining, and `jsgen`
handles the actual inlining.

#### Current Exception Tracking

When control-flow leaves an `except` handler, the `prevJSError` global
(whose content is returned by `getCurrentException`) needs to be reset
to the exception caught by the enclosing handler, something that was
previously missing for the most part.

The exception is restored:
* at the end of a block when the block is broken out of from within a
  `catch`
* at the start of a `finally` clause intercepting a `break` from within
  a `catch`
* at the start of a `finally` clause when it's entered due to an
  exception

If there's no enclosing `catch` within a procedure when the current
exception has to be restored, the value of `lastJSError` is captured at
procedure entry. Tracking of the current exception still doesn't
include in-flight exceptions, so `tfinally6.nim` still fails.

#### Unhandled Exception Reporting

Instead of incrementing, at run-time, a global counter (`excHandler`)
before entering a `try` and decrementing it again on exit, and then
reporting unhandled exceptions when throwing, the module-init
procedures (i.e., top-level code) are wrapped in a `try`/`catch` that
reports the escaping exceptions.

This is simpler to implement and should also have less run-time
overhead.

#### Throwing with `asm`

`asm` and `emit` statements are considered not to raise, and thus they
don't prevent `except` and `finally` sections from being eliminated by
unreachable code elimination.

The `tnativeexc` test is updated to not rely on undefined behaviour.

### Unreachable Code Elimination

Labels that mark the exit of case statements are now eliminated too, if
structured control-flow doesn't leave the case statement (i.e., each
branch ends in a noreturn-like statement).

zerbina committed:
jsbackend: enable new-style CGIR

zerbina committed:
jsgen: account for checked calls

They have a label/target-list in the last slot, which needs to be
skipped when counting or iterating over the arguments.

zerbina committed:
introduce `jsflow` module

The module provides the `toStructureList` procedure, which translates
the goto-based control-flow representation of CGIR to JavaScript
control-flow constructs.

Since JavaScript has very high-level constructs - that also match their
NimSkull counterparts fairly well -, the translation is much simpler
than that used by `cgen`.

zerbina committed:
jsgen: integrate `jsflow`

`genStmts` takes the statement list and creates the JavaScript control-
flow representation from it, the latter which then drives code
generation for statements. Proper indentation (which is only cosmetic)
is mostly handled by `genStmts` through the new `startBlock` and
`endBlock` templates.

Tracking of the current exception (stored in `lastJSError`) doesn't
work yet.

zerbina committed:
jsgen: remove obsolete routines and fields

The codegen routines for the old CGIR control-flow constructs are
obsolete and thus removed. Block tracking and injection of the
`BeforeRet_` label is obsolete too.

zerbina committed:
jsflow: properly handle jump-to-finally

A `break <label>` was emitted, but there's no label for a finally, so
the JS parser rightfully complains. A `return` is now emitted in these
cases instead.

zerbina committed:
jsgen: implement proper current exception tracking

The current exception (stored in `lastJSError`) is now properly updated
after leaving an exception handler. This is achieved by setting
`lastJSError` to the exception of the immediately enclosing `catch`
after landing at a join point that's reachable through a jump crossing
an `except` boundary.

zerbina committed:
cgirgen: omit unnecessary join at end of `case`

`jsflow` expects all joins to have some gotos that target them
(otherwise it crashes), which was not always the case.

To not crash with an index defect when there's a bug with `cgirgen`
(or `jsflow`), `jsflow` now asserts that all unstructured join points
have a corresponding opening.

zerbina committed:
jsgen: restore reporting of unhandled exceptions

Instead of keeping track of the enclosing `try` blocks at run-time, for
the purpose of knowing when to report an unhandled exception, module-
level code is now wrapped in a `try`/`catch` that reports escaping
exceptions. This is both simpler to implement *and* has less run-time
overhead.

The `excHandler` global is removed.

zerbina committed:
tests: enable multiple tests for the JavaScript target

All tests previously failing due to improper "current exception"
tracking (outside of `finally`) now work.

zerbina committed:
jsgen: restore catching of native exceptions

zerbina committed:
tests: update `tnativeexc`

An `asm` statement (or `emit` statement, for that matter) is considered
not being able to raise, which now leads to the exception handlers
being eliminated by unreachable code removal.

Imported procedures are considered as potentially raising exceptions.

zerbina committed:
jsgen, jsflow: inline case dispatcher bodies

To eliminate very deep block statement nesting for large case
statements, the code generator now tries to inline the code following
jump targets directly within the generated switch-case statements.

Implementation-wise, `jsflow` uses a simple heuristic that applies to
most, if not all, to compute the set of blocks that can be eliminated.
`jsgen` then skips the code following eliminated blocks, and instead
generates it directly within the switch-case statements. Unfortunately,
this re-introduces input- dependent call recursion into the code
generator.

zerbina committed:
jsflow: add missing `of` branch labels

zerbina commented:
`jsgen` now tries to inline the jump target of case dispatcher statements into the switch-case statement. This fixes node.js failing to compile the megatest (`thtmlparser.nim` is the offender here). As a side note, I've also tried to run the code in FireFox (SpiderMonkey engine), and there it works.

The heuristic that's currently in use is not correct however, and some case dispatcher shapes (none of which are possible at the moment) would cause incorrect code generation.

zerbina committed:
jsgen: consider jumps of checked calls

Jumps of checked calls (calls that can potentially raise an exception)
are now also handled. Since the jump is not guaranteed, sections
tentatively disabled prior to the call are re-enabled again directly
after the call.

zerbina committed:
jsflow: remove "needs recover" tracking

Where the current exception must be restored is now a local analysis
part of the `handleJump` analysis. This reduces the amount of global
and disjoint state a bit.

zerbina committed:
jsgen: ensure correct try/finally/catch control-flow

It's possible for the JavaScript `try`s produced for some MIR/CGIR
to not exhibit the required behaviour, as, for example, finally clauses
are entered when they shouldn't be (can happen when a goto jumps over
a finally section).

To prevent this, intermixed with code generation, all jump actions are
analysed for whether the path matches that taken at run-time. If it
wouldn't `catch` and `finally` clauses are flagged as having to be
guarded by boolean flag, which is then toggled prior to a problematic
jump.

Checked calls aren't handled yet.

zerbina committed:
jsflow: don't inline fall-through regions

Regions of JavaScript code that don't end in a terminator (break,
raise, or return) cannot be inlined into case statements, as that would
result in the wrong behaviour.

zerbina commented:
Okay, that should be it. `catch` and `finally` clauses for which it is necessary can now be disabled at run-time, resolving problems a mismatch between expected and actual control-flow. For example, the following MIR (CGIR at the moment):
```
if ...
  goto [L1, L2]
end
goto L2
finally L1:
  ...
  continue
L2:
...
```
translates to:
```js
L2: {
  try {
    if (...) {
      break L2;    
    }
    break L2;
  } finally {
    ...
  }
}
```
The second `break` would also visit the `finally`, even though the original goto explicitly states that it should not.

While CGIR such as the above is not possible at the moment, except handlers are not merged into one at the moment, so:
```nim
try:
  raise ...
except A as e:
  ...
except:
  ...
```
is currently translated into:
```js
try {
  try {
    throw ...
  } catch (Exception1_) {
    // rethrow if no type match
  }
} catch (Exception1_) {
  ...
}
```
where it has to be made ensure that exceptions raised from within the first exception handler are not handled by the outer `try`.

zerbina committed:
jsgen: fix two bugs with "section enabled" management

* the `Enabled` local being undefined wasn't considered (undefined is
  now treated as true)
* re-using the same `Enabled` name multiple times can lead to improper
  behaviour in edge cases. The name is derived from the sections'
  unique label ID now

zerbina commented:
The changes do increase the complexity of the code generator quite a bit. I think at least part of it is due to the ad-hoc nature of the change: `jsgen` was not designed with this kind of translation in mind -- rewriting `jsgen` from scratch against the MIR *should* reduce some of the complexity. However, I think that the larger part of the reason is that going from low-level, goto-using code to code with breaks, blocks, and try/catch/finally has just some inherent complexity.

I experimented with different approaches, such as having `jsflow` emit a more elaborate IR (akin to `ccgflow`), or performing the switch-case body inlining in `jsflow`, but the solution I converged on ultimately felt like the least complex.

zerbina committed:
jsflow: consider `return`s during inlineability detection

zerbina committed:
jsgen: small simplification

zerbina committed:
improve comments, remove stale ones, fix typos

zerbina committed:
jsgen: only emit the definition enable flags once

It's not necessary for behaviour, but emitting less code is generally
better.

zerbina committed:
jsgen: fix yet another enable-flag bug

Consider:
```js
while (true) {
  L1: {
    try {
      if (...) {
        var Enabled1_ = true;
        Enabled1_ = false;
        break L1;
      } else {
        break L1;
      }
    } finally {
      if (Enabled1_ === undefined || Enabled_1) {
        ...
      }
    }
  }
}
```
This code won't behave as expected, as the second break doesn't execute
the finally. The flag needs to be set back to true on exit of the
section.

The naming change from earlier could be reverted again, but using
distinct names makes the code a bit clearer, so it's kept.

zerbina committed:
jsgen: fix missing current exception restoration

A small oversight. Since the current exception is restored when
*landing* (instead of when *leaving*), it's possible that an exception
handler within a procedure is exited through raising another exception,
but without there being an intercepting finally that restores the
exception. Therefore, all exception control-flow leaving procedures
needs to be treated as requiring a restore on landing.

It's not possible to add a test for this behaviour, since it's at odds
with the behaviour of the C and VM backend.

saem commented on compiler/backend/jsgen.nim:
```diff
@@ -1227,7 +1272,7 @@ proc genArgs(p: PProc, n: CgNode, r: var TCompRes; start=1) =
   var emitted = start-1
 
-  for i in start..<n.len:
+  for i in start..<(1 + numArgs(n)):
```
```suggestion
  for i in start..numArgs(n):
```
since `<` is basically a `- 1`, I'm assuming this simplification works

zerbina replied:
I've replaced the `(1 + numArgs(n))` expressions with a call to the new `callLen` procedure, which does the same thing but should convey the intention better.

saem commented on compiler/backend/jsgen.nim:
```diff
@@ -1247,11 +1292,11 @@ proc genArgs(p: PProc, n: CgNode, r: var TCompRes; start=1) =
 proc genOtherArg(p: PProc; n: CgNode; i: int; typ: PType;
                  generated: var int; r: var TCompRes) =
-  if i >= n.len:
+  if i >= numArgs(n) + 1:
```
```suggestion
  if i > numArgs(n):
```

I believe this simplification is still equivalent.

saem commented on compiler/backend/jsgen.nim:
```diff
@@ -1274,7 +1319,7 @@ proc genPatternCall(p: PProc; n: CgNode; pat: string; typ: PType;
     of '@':
       var generated = 0
-      for k in j..<n.len:
+      for k in j..<(1 + numArgs(n)):
```
```suggestion
      for k in j..numArgs(n):
```
same as before

saem commented on compiler/backend/jsgen.nim:
```diff
@@ -1341,7 +1386,7 @@ proc genEcho(p: PProc, n: CgNode, r: var TCompRes) =
   r.res.add("rawEcho(")
   # the first argument is a literal type that we don't need
-  for i in 2..<n.len:
+  for i in 2..<(1 + numArgs(n)):
```
```suggestion
  for i in 2..<numArgs(n):
```

saem commented on compiler/backend/jsgen.nim:
```diff
@@ -2087,15 +2126,23 @@ proc genProcBody(p: PProc, prc: PSym): Rope =
+  elif sfModuleInit in prc.flags:
+    # report an unhandled exception when a |NimSkull| exception escapes
+    # module-level code
+    # XXX: this is common logic across all backends; it needs to be handled at
```
Ah fun! This also provides some clues as to how we might handle `defect` -> `exception` conversion for crossing package bounardies.

saem commented on compiler/backend/jsflow.nim:
```diff
@@ -0,0 +1,360 @@
+const
+  Terminators = {stkReturn, stkTerminator}
+
+func finalTarget*(n: CgNode): CgNode =
```
```suggestion
func finalTarget*(n: CgNode): CgNode =
  ## given a label or target list, retrieves the target
```

zerbina replied:
Thanks, I missed that one.

saem reviewed (approved):
I'm under the weather, so I don't think I was all that thorough in my review, but the conversion looks good.

Varriount commented:
Why does translation of switch/case cause nesting? Wouldn't something like 
```nim
case direction
of North:
  ...
of South:
  ...
of East:
  ...
of West:
  ...
```
be translated to (roughly)
```javascript
switch (direction) {
  case North:
    ...
    break;
  case South:
    ...
    break;
  case East
    ...
    break;
  case West
    ...
    break;
}
```

Or does the goto-based IR make that impossible?

zerbina committed:
jsflow: document `finalTarget`

zerbina committed:
jsgen: replace `numArgs(n) + 1` with `callLen`

The new `callLen` utility procedure doesn't evaluates to the same
thing, but it conveys the intention much better.

Varriount commented on compiler/backend/jsflow.nim:
```diff
@@ -0,0 +1,360 @@
+      discard "not relevant"
+
+  # the list of openings and closing produced by the first pass will in most
+  # cases not be valid JavaScript code. We have to "solve" therepresentation
```
```suggestion
  # cases not be valid JavaScript code. We have to "solve" the representation
```

zerbina commented:
> Why does translation of switch/case cause nesting? Wouldn't something like
> 
> be translated to (roughly)

That's how the translation works now, yes. In the MIR (CGIR, at the moment), a `case` statement is an exhaustive dispatcher that works like this:
```nim
case value
of 1: goto L1
of 2: goto L2
of 3: goto L3
L1:
# branch 1
goto CaseExit
L2:
# branch 2
goto CaseExit
L3:
# branch 3
goto CaseExit
CaseExit:
```

Gotos are translated to `break` statement, so the naive translation (the one used prior to [this](https://github.com/nim-works/nimskull/pull/1230/commits/180bf94da2c4136ccfaa3062acfa9edf49768931) commit) results in the following JavaScript output:
```js
CaseExit: {
  L3: {
    L2: {
      L1: {
        switch (value) {
        case 0: break L1;
        case 1: break L2;
        case 2: break L3;
        }
      }
      ...
      break CaseExit;
    }
    ...
  }
  ...
}
```

zerbina committed:
fix a typo

Co-authored-by: Clay Sweetser <Varriount@users.noreply.github.com>

Clyybber commented on compiler/backend/jsflow.nim:
```diff
@@ -0,0 +1,361 @@
+    stkCatch   ## a 'catch' clause of a 'try' statement
+    stkFinally ## a 'finally' clause of a 'try' statement
+
+    stkEnd     ## end of a catch, finally, or block
```
AFAICT this also marks the end of a `stkStructStart`?

zerbina replied:
Good catch, it does. I'll update the comment.

zerbina committed:
jsflow: clarify `stkEnd` documentation

zerbina commented:
/merge

Thank you for the reviews, everyone!
